---
title:  Word network of Bioconductor packages 
author: maria
date: '2021-01-30'
categories:
  - R
tags:
  - r
  - BiocPkgTools
  - Bioconductor
  - Tidytext
  - widyr
  - ggraph
slug:  network visualization of Bioconductor packages 
subtitle: Exploration of the connections between Bioconductor packages
summary: Understanding how Bioconductor packages are connected between each other using metadata.
lastmod: '2021-01-30'
featured: yes
draft: no
image:
  placement: 1
  caption: 'Word network in Bioconductor packages titles'
  focal_point: Center
  preview_only: no
output:
  blogdown::html_page:
    toc: yes
    number_sections: no
    toc_depth: 1
links:
  - icon: 
    icon_pack: fab
    name: Bioconductor metadata available at BiocPkgTools 
    url: https://seandavi.github.io/BiocPkgTools/
---

<script src="{{< blogdown/postref >}}index_files/header-attrs/header-attrs.js"></script>

<div id="TOC">
<ul>
<li><a href="#motivation">Motivation</a></li>
<li><a href="#statistics-of-bioconductor-downloads">Statistics of Bioconductor downloads</a></li>
<li><a href="#full-details-of-bioconductor-packages">Full details of Bioconductor packages</a></li>
<li><a href="#word-network-of-bioconductor-packages">Word network of Bioconductor packages</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>
</div>

<pre class="r"><code>library(BiocPkgTools)
library(tidyverse)
library(tidytext)
library(widyr)
library(igraph)
library(ggraph)
library(lubridate)
library(emo)</code></pre>
<div id="motivation" class="section level1">
<h1>Motivation</h1>
<p>Bioconductor has a total of 5796 at the present day 2021-01-31. Therefore, navigating across Bioconductor packages can be a daunting experience. Luckily, <a href="https://seandavi.github.io/BiocPkgTools/">BiocPkgTools</a> offers a rich ecosystem of metadata around Bioconductor packages üìù.</p>
</div>
<div id="statistics-of-bioconductor-downloads" class="section level1">
<h1>Statistics of Bioconductor downloads</h1>
<p>We can get a tidy data.frame with download stats for all packages using the function <code>biocDownloadStats</code>.</p>
<pre class="r"><code>#  Getting a tidy tibble summarizing monthly download statistics 
bio_download_stats &lt;- biocDownloadStats()</code></pre>
<pre class="r"><code>bio_download_stats %&gt;% 
  head(13)</code></pre>
<pre><code>## # A tibble: 13 x 7
##    Package  Year Month Nb_of_distinct_IPs Nb_of_downloads repo     Date      
##    &lt;chr&gt;   &lt;int&gt; &lt;chr&gt;              &lt;int&gt;           &lt;int&gt; &lt;chr&gt;    &lt;date&gt;    
##  1 ABarray  2021 Jan                   54             114 Software 2021-01-01
##  2 ABarray  2021 Feb                    0               0 Software 2021-02-01
##  3 ABarray  2021 Mar                    0               0 Software 2021-03-01
##  4 ABarray  2021 Apr                    0               0 Software 2021-04-01
##  5 ABarray  2021 May                    0               0 Software 2021-05-01
##  6 ABarray  2021 Jun                    0               0 Software 2021-06-01
##  7 ABarray  2021 Jul                    0               0 Software 2021-07-01
##  8 ABarray  2021 Aug                    0               0 Software 2021-08-01
##  9 ABarray  2021 Sep                    0               0 Software 2021-09-01
## 10 ABarray  2021 Oct                    0               0 Software 2021-10-01
## 11 ABarray  2021 Nov                    0               0 Software 2021-11-01
## 12 ABarray  2021 Dec                    0               0 Software 2021-12-01
## 13 ABarray  2021 all                   54             114 Software NA</code></pre>
<p>As we see observations for all the months of the year are generated once that the year starts (download values for events in the future are filled up with <em>0</em>). Also note that there is a summary statistic for month called <code>all</code> embedded inside the tibble, and the <code>Date</code> value for that observation is NA (this would makes group by date very convenient).</p>
<p>This tibble contains information about packages that expands from 2009 to 2021. There are 3 categories of packages, with the total number of package per category as follows:</p>
<pre class="r"><code>bio_download_stats %&gt;% 
  distinct(Package, repo) %&gt;%
  count(repo) %&gt;%
  knitr::kable()</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">repo</th>
<th align="right">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">AnnotationData</td>
<td align="right">2659</td>
</tr>
<tr class="even">
<td align="left">ExperimentData</td>
<td align="right">821</td>
</tr>
<tr class="odd">
<td align="left">Software</td>
<td align="right">2316</td>
</tr>
</tbody>
</table>
</div>
<div id="full-details-of-bioconductor-packages" class="section level1">
<h1>Full details of Bioconductor packages</h1>
<p>The complete information for the packages as described in the <code>DESCRIPTION</code> file can be obtained with <code>biocPkgList</code>.</p>
<pre class="r"><code>bpi = biocPkgList()
colnames(bpi)</code></pre>
<pre><code>##  [1] &quot;Package&quot;               &quot;Version&quot;               &quot;Depends&quot;              
##  [4] &quot;Suggests&quot;              &quot;License&quot;               &quot;MD5sum&quot;               
##  [7] &quot;NeedsCompilation&quot;      &quot;Title&quot;                 &quot;Description&quot;          
## [10] &quot;biocViews&quot;             &quot;Author&quot;                &quot;Maintainer&quot;           
## [13] &quot;git_url&quot;               &quot;git_branch&quot;            &quot;git_last_commit&quot;      
## [16] &quot;git_last_commit_date&quot;  &quot;Date/Publication&quot;      &quot;source.ver&quot;           
## [19] &quot;win.binary.ver&quot;        &quot;mac.binary.ver&quot;        &quot;vignettes&quot;            
## [22] &quot;vignetteTitles&quot;        &quot;hasREADME&quot;             &quot;hasNEWS&quot;              
## [25] &quot;hasINSTALL&quot;            &quot;hasLICENSE&quot;            &quot;Rfiles&quot;               
## [28] &quot;dependencyCount&quot;       &quot;Imports&quot;               &quot;Enhances&quot;             
## [31] &quot;dependsOnMe&quot;           &quot;VignetteBuilder&quot;       &quot;suggestsMe&quot;           
## [34] &quot;LinkingTo&quot;             &quot;Archs&quot;                 &quot;URL&quot;                  
## [37] &quot;SystemRequirements&quot;    &quot;BugReports&quot;            &quot;importsMe&quot;            
## [40] &quot;Video&quot;                 &quot;linksToMe&quot;             &quot;OS_type&quot;              
## [43] &quot;PackageStatus&quot;         &quot;License_restricts_use&quot; &quot;License_is_FOSS&quot;      
## [46] &quot;organism&quot;</code></pre>
<p>There is lots of information in here. We could use this metadata information to understand the connections between packages.</p>
</div>
<div id="word-network-of-bioconductor-packages" class="section level1">
<h1>Word network of Bioconductor packages</h1>
<p>The most informative variables about the packages are <code>Title</code> and <code>Description</code> so we can explore the connections between packages doing some <strong>text mining</strong> using a <a href="https://www.tidytextmining.com/index.html">Tidytext</a> approach.</p>
<p>To prepare our dataset we need to initially tokenize the text. The Wikipedia definition for <a href="https://en.wikipedia.org/wiki/Lexical_analysis#Tokenization">tokenization</a> on lexical analysis is as follows:</p>
<blockquote>
<p><strong>Tokenization is the process of demarcating and possibly classifying sections of a string of input characters</strong></p>
</blockquote>
<p>The <em>sections</em> can be words, sentence, ngram or chapter (for example if analysis a book). In this case we are gonna break down package Titles or Description into words using the function <code>unnest_tokens</code>.
In addition, we can remove <a href="https://en.wikipedia.org/wiki/Stop_word">stop words</a> (included in the Tidytext dataset).</p>
<pre class="r"><code>bpi_title &lt;- bpi %&gt;% 
  dplyr::select(Package, Title) %&gt;%
  unnest_tokens(word, Title) %&gt;% 
  anti_join(stop_words)

bpi_description &lt;- bpi %&gt;%
  dplyr::select(Package, Description) %&gt;%
  unnest_tokens(word, Description) %&gt;%
  anti_join(stop_words)</code></pre>
<p>Note that the number of words from Title is 11932 and the number
of words from Description is 59370, so package Descriptions
contain on average 5 times the words of package Titles.</p>
<p>We can have a look at how the tokenised titles for each package look like:</p>
<pre class="r"><code>bpi_title</code></pre>
<pre><code>## # A tibble: 11,932 x 2
##    Package word      
##    &lt;chr&gt;   &lt;chr&gt;     
##  1 a4      automated 
##  2 a4      affymetrix
##  3 a4      array     
##  4 a4      analysis  
##  5 a4      umbrella  
##  6 a4      package   
##  7 a4Base  automated 
##  8 a4Base  affymetrix
##  9 a4Base  array     
## 10 a4Base  analysis  
## # ‚Ä¶ with 11,922 more rows</code></pre>
<p>Them, we can use <code>pairwise_count</code> from the <a href="https://cran.r-project.org/web/packages/widyr/vignettes/intro.html">widyr</a> package to count how many times each pair of words occurs together in the package Title. This function works as a mutate in that it takes the variables to compare and returns a tibble with the pairwise columns and an extra column called <code>n</code> containing the number of words co-occurrences. I think this function is very sweet üçØ!</p>
<pre class="r"><code>bpi_titlepairs &lt;- bpi_title %&gt;%
pairwise_count(Package, word, sort = TRUE, upper = FALSE)

bpi_desciptionpairs &lt;- bpi_description %&gt;%
pairwise_count(Package, word, sort = TRUE, upper = FALSE)</code></pre>
<p>This data is ready for visualization of network of co-occurring words in package Titles. We can use the <code>ggraph</code> package for visualizing this network. We are going to represent just the top co-occurring words, or otherwise we get a very populated network which is impossible to read.</p>
<pre class="r"><code>set.seed(1234)
bpi_titlepairs %&gt;%
  filter(n &gt;= 6) %&gt;%
  graph_from_data_frame() %&gt;%
  ggraph(layout = &quot;fr&quot;) +
  geom_edge_link(aes(edge_alpha = n, edge_width = n), edge_colour = &quot;purple&quot;) +
  geom_node_point(size = 5) +
  geom_node_text(aes(label = name), repel = TRUE, 
                 point.padding = unit(0.2, &quot;lines&quot;)) +
  theme_void()+
  theme(legend.position=&quot;none&quot;)+
  labs(title = &quot;  Number of word co-ocurrences in packages titles&quot;)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-10"></span>
<img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-10-1.png" alt="Word network in Bioconductor packages Titles" width="672" />
<p class="caption">
Figure 1: Word network in Bioconductor packages Titles
</p>
</div>
<p>We see some clear and logical clustering of packages in this network.For example, DESEq and DESeq2 packages cluster together, as one would expect since they DESeq2 is the successor of DESeq. There are other obvious connections such as MSstatsTMTPTM and MSstatsTMTP since the former has added functionality to analyse PTMs on TMT shotgun mass spectrometry-based proteomic experiments.
There is a big cluster on the bottom left corner with packages to analyse RNASeq and single cell RNASeq.</p>
<p>What about the network build from words of the Description?</p>
<pre class="r"><code>set.seed(1234)
bpi_desciptionpairs %&gt;%
  filter(n &gt;= 15) %&gt;%
  graph_from_data_frame() %&gt;%
  ggraph(layout = &quot;fr&quot;) +
  geom_edge_link(aes(edge_alpha = n, edge_width = n), edge_colour = &quot;orange&quot;) +
  geom_node_point(size = 5) +
  geom_node_text(aes(label = name), repel = TRUE, 
                 point.padding = unit(0.2, &quot;lines&quot;)) +
  theme_void()+
  theme(legend.position=&quot;none&quot;)+
  labs(title = &quot;Number of word co-ocurrences in packages Description&quot;)</code></pre>
<div class="figure"><span id="fig:unnamed-chunk-11"></span>
<img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-11-1.png" alt="Word network in Bioconductor packages Description" width="672" />
<p class="caption">
Figure 2: Word network in Bioconductor packages Description
</p>
</div>
<p>We see more connections here, and some of the relationships are still obvious (e.g HiCcompare and multiHiCcompare, anota and anota2seq, AnnotationHub and ExperimentHub). This network is richer, and one would have to dive a bit deeper to get a better sense of this network.</p>
</div>
<div id="conclusions" class="section level1">
<h1>Conclusions</h1>
<p>Text mining of Bioconductor packages metadata is a straight forward visual way to understand the relationships between packages. One could go beyond this and for example find words that are
especially important across package Descriptions by calculating <a href="https://www.tidytextmining.com/tfidf.html#tfidf">tf-idf statistic</a>. One could also set up a GitHub Action executed as a CRON job to get updates periodically. This could turn into a challenge for <a href="https://kevinrue.github.io/BiocChallenges/index.html">BiocChallenges</a>. This post was inspired by <a href="https://www.tidytextmining.com/nasa.html">Chapter 8</a> of the Tidytext book and <a href="https://kevinrue.github.io/post/biocpkgtools/">BiocRoulette</a>.</p>
</div>
